b0 bootloader: new optional mcuboot trailer to support upgradable mcuboot

mads-bn opened the pull request:
b0 bootloader: new optional image trailer to support upgradable mcuboot

This feature is always enabled but NOP if magic header not detected.
The new optional image trailer in mcuboot s0 and s1 partitions are located in the
last 16 bytes of each partition.
This feature also require a similar extension of mcuboot.

b0 validates and selects which mcuboot slot to boot from (supports fallback, if unconfirmed):
1) mcuboot erase flash sector for mcuboot_image_trailer
2) mcuboot writes new image to flash (e.g. "mcumgr image upload")
3) mcuboot writes mcuboot_image_trailer with value mcuboot_image_trailer_init as final step of image upload
4) device is rebooted via "mcumgr reboot" or external reset
5) b0 checks if one or both mcuboot slots have a trailer.magic, if not normal boot proceeds
6) b0 updates mcuboot_image_trailer.status appropriately:
   From TESTING to BOOTING (if any TESTING)
   From BOOTING to INACTIVE (if any BOOTING)
7) b0 gives a slot 1st priority if its trailer.status=TESTING
   otherwise it prioritizes trailer.status=ACTIVE
8) mcuboot changes trailer.status of booted image to ACTIVE if confirmed by "mcumgr image confirm"
   and simultaneously marks the not-booted image as INACTIVE

NOTE: if any image fails to validate, the other bank will still be attempted as well.
Hence b0 can "fallback" to an image otherwise marked as INACTIVE.

Commit:
b0 bootloader: new optional image trailer to support upgradable mcuboot

This feature is always enabled but NOP if magic header not deteted.
The new optional image trailer in mcuboot s0 and s1 partitions are located in the
last 16 bytes of each partition.

b0 validates and selects which mcuboot slot to boot from (supports fallback, if unconfirmed):
1) mcuboot erase flash sector for mcuboot_image_trailer
2) mcuboot writes new image to flash (e.g. "mcumgr image upload")
3) mcuboot mcuboot_image_trailer with mcuboot_image_trailer_init
4) device is rebooted via "mcumgr reboot" or external reset
5) b0 checks if both mcuboot slots have a valid mcuboot_image_trailer.magic, if not normal boot proceeds
6) b0 changes mcuboot_image_trailer.status from TESTING to BOOTING (if any TESTING)
7) b0 gives slot 1st priority if its trailer.status=TESTING/ACTIVE
8) mcuboot changes trailer.status from BOOTING to ACTIVE if confirmed by "mcumgr image confirm"
   mcuboot also mark passive slot INVALID, to avoid booting in that (except fallback)

Emil-Juhl commented:
Nitpick for commit:
> The new optional image trailer in mcuboot s0 and s1 partitions are located in the
last 16 bytes of each partition.
>

It's actually the last 16 bytes of the last sector/page of the partition. As we've discussed, currently this means that the image trailer is actually located 512 bytes (plus the size of the trailer, so 528 bytes) from the partition end address.

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
 #define TYPE_COUNTERS 1 /* Type referring to counter collection. */
 #define COUNTER_DESC_VERSION 1 /* Counter description value for firmware version. */
 
+
```
```suggestion
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+ * optional trailer at the end of of flash partitions for mcuboot slot0 and slot1
+ * 1) mcuboot erase flash sector for mcuboot_image_trailer
+ * 2) mcuboot writes new image to flash (e.g. "mcumgr image upload")
+ * 3) mcuboot mcuboot_image_trailer with mcuboot_image_trailer_init
```
```suggestion
 * 3) mcuboot writes mcuboot_image_trailer with value mcuboot_image_trailer_init as final step of image upload
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+ * 2) mcuboot writes new image to flash (e.g. "mcumgr image upload")
+ * 3) mcuboot mcuboot_image_trailer with mcuboot_image_trailer_init
+ * 4) device is rebooted via "mcumgr reboot" or external reset
+ * 5) b0 checks if both mcuboot slots have a valid mcuboot_image_trailer.magic, if not normal boot proceeds
```
I think it should be OK if only one of them has the trailer? Otherwise I think we might have issues on the first flash unless we update our .bin's somehow to include a basic image_trailer as well

mads-bn replied:
Now:
> * 5) b0 checks if one or both mcuboot slots have a trailer.magic, if not normal boot proceeds

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+ * 3) mcuboot mcuboot_image_trailer with mcuboot_image_trailer_init
+ * 4) device is rebooted via "mcumgr reboot" or external reset
+ * 5) b0 checks if both mcuboot slots have a valid mcuboot_image_trailer.magic, if not normal boot proceeds
+ * 6) b0 changes mcuboot_image_trailer.status from TESTING to BOOTING (if any TESTING)
```
```suggestion
 * 6) b0 updates mcuboot_image_trailer.status appropriately:
            From TESTING to BOOTING (if any TESTING)
            From BOOTING to INACTIVE (if any BOOTING)
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+ * 4) device is rebooted via "mcumgr reboot" or external reset
+ * 5) b0 checks if both mcuboot slots have a valid mcuboot_image_trailer.magic, if not normal boot proceeds
+ * 6) b0 changes mcuboot_image_trailer.status from TESTING to BOOTING (if any TESTING)
+ * 7) b0 gives slot 1st priority if its trailer.status=TESTING/ACTIVE
```
```suggestion
 * 7) b0 gives a slot 1st priority if its trailer.status=TESTING
      otherwise it prioritizes trailer.status=ACTIVE
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+ * 5) b0 checks if both mcuboot slots have a valid mcuboot_image_trailer.magic, if not normal boot proceeds
+ * 6) b0 changes mcuboot_image_trailer.status from TESTING to BOOTING (if any TESTING)
+ * 7) b0 gives slot 1st priority if its trailer.status=TESTING/ACTIVE
+ * 8) mcuboot changes trailer.status from BOOTING to ACTIVE if confirmed by "mcumgr image confirm"
```
```suggestion
 * 8) mcuboot changes trailer.status of booted image to ACTIVE if confirmed by "mcumgr image confirm"
      and simultaneously marks the not-booted image as INACTIVE
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+ * 7) b0 gives slot 1st priority if its trailer.status=TESTING/ACTIVE
+ * 8) mcuboot changes trailer.status from BOOTING to ACTIVE if confirmed by "mcumgr image confirm"
+ *    mcuboot also mark passive slot INVALID, to avoid booting in that (except fallback)
+ */
```
```suggestion
 * NOTE: if any image fails to validate, the other bank will still be attempted as well. Hence
 *        b0 can "fallback" to an image otherwise marked as INACTIVE.
 */
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+ */
+
+/* Note1: interface cloned by mcuboot/boot/boot_serial/src/boot_serial.c */
+/* Note2: these constants rely on bits in flash can be changed from 1 to 0 without erasing */
```
:fire: important comment!

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+
+typedef struct {
+    uint32_t status; /* mcuboot writes TESTING or ACTIVE, b0 writes BOOTING if TESTING found */
+    uint32_t magic[MCUBOOT_IMAGE_TRAILER_MAGIC_SZ]; /* mcuboot writs*/
```
```suggestion
    uint32_t magic[MCUBOOT_IMAGE_TRAILER_MAGIC_SZ]; /* mcuboot writes*/
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+
+slot_priority slot_priority_from_image_trailers(void)
+{
+	int result = SLOT_PRIORITY_S0;
```
```suggestion
        /* preserve behavior of b0 if no magic headers are present, by "prioritizing" S0 as default */
	int result = SLOT_PRIORITY_S0;
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+	int result = SLOT_PRIORITY_S0;
+	uint32_t s0_addr = s0_address_read();
+	uint32_t s1_addr = s1_address_read();
+	if (s0_addr == PM_S0_OFFSET && s1_addr == PM_S1_OFFSET) {
```
```suggestion
        /* sanity check that the addresses from `provisioning` and the PM config match */
	if (s0_addr == PM_S0_OFFSET && s1_addr == PM_S1_OFFSET) {
```

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+	uint32_t s0_addr = s0_address_read();
+	uint32_t s1_addr = s1_address_read();
+	if (s0_addr == PM_S0_OFFSET && s1_addr == PM_S1_OFFSET) {
+		const mcuboot_image_trailer* s0 = (mcuboot_image_trailer*)(s0_addr + PM_S0_SIZE - sizeof(mcuboot_image_trailer));
+		const mcuboot_image_trailer* s1 = (mcuboot_image_trailer*)(s1_addr + PM_S1_SIZE - sizeof(mcuboot_image_trailer));
```
We need to fix the read address here. They don't align with the current mcuboot patches. Essentially you lack another `- 512` in there

mads-bn replied:
I hope we can drop this requirement after PM fixed -  maybe add a compile time assert in mcuboot, which otherwise could erase wrong section.

Emil-Juhl replied:
Yeah, ignore this one.

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +39,100 @@ struct counter_collection {
+		const mcuboot_image_trailer* s1 = (mcuboot_image_trailer*)(s1_addr + PM_S1_SIZE - sizeof(mcuboot_image_trailer));
+		uint32_t s0_status = check_and_fix_image_trailer("slot0", s0);
+		uint32_t s1_status = check_and_fix_image_trailer("slot1", s1);
+		if (s0_status == MCUBOOT_IMAGE_TRAILER_STATUS_INACTIVE || s1_status == MCUBOOT_IMAGE_TRAILER_STATUS_TESTING) {
+			if (s1_status != MCUBOOT_IMAGE_TRAILER_STATUS_INACTIVE)
```
```suggestion
		if (s1_status == MCUBOOT_IMAGE_TRAILER_STATUS_ACTIVE || s1_status == MCUBOOT_IMAGE_TRAILER_STATUS_TESTING) {
```
I _think_ you could end up looping (if we didn't also change `BOOTING -> INACTIVE` a test partition. Either way I believe the logic is equivalent to what we need and maybe easier to read.
Only prioritize S1 _if_ ACTIVE _or_ TESTING

Let me know, if you disagree. I don't doubt that the current logic works out!

mads-bn replied:
I puzzled with this part but I think S1 could be active while S0=TESTING ?
But I agree if we can simplify it it would be easier to read ...

Emil-Juhl replied:
Ah, that's true. That case needs to be handled as well!

mads-bn replied:
Will try test this:
<pre>
		switch (s1_status) {
		case MCUBOOT_IMAGE_TRAILER_STATUS_TESTING:
		case MCUBOOT_IMAGE_TRAILER_STATUS_BOOTING:
		case MCUBOOT_IMAGE_TRAILER_STATUS_ACTIVE:
			switch (s0_status) {
			case MCUBOOT_IMAGE_TRAILER_STATUS_TESTING:
			case MCUBOOT_IMAGE_TRAILER_STATUS_BOOTING:
				break; /* s0 has priority */
			default:
				result = SLOT_PRIORITY_S1;
			break;
		}
</pre>

Emil-Juhl replied:
Should the BOOTING state also be included here? Won't that mean, that we essentially get two attempts when switching bank? (Although that could be practical in some cases to e.g. test that also the FEP application starts OK with the new mcuboot)

mads-bn replied:
we will keep this - at least for now

Emil-Juhl commented on samples/bootloader/src/main.c:
```diff
@@ -92,8 +92,12 @@ void main(void)
+	slot_priority priority = slot_priority_from_image_trailers();
 
-	if (!s1_info || (s0_info->version >= s1_info->version)) {
+	if (!s1_info && (s0_info->version > s1_info->version)) {
```
You're gonna access a bad pointer here.
I think, if we really want to preserve behavior here, we might actually need an "undecided" priority to check against.
```
if (priority == NONE) {
    if (!s1_info || (s0_info->version >= s1_info->version))
```

Alternatively actually you might just need to not flip the `||` to `&&` and keep the remaining logic where you substitute `>=` for `>`
This does mean that we'll _always_ prioritize the bigger version of S0 though. So then you probably need an equivalent prioritization for if the version of S1 is greater than S0.

Maybe
```
if (s1_info && (s0_info->version != s1_info->version)) {
    if (s0_info->version > s1_info->version)
        priority = SLOT_PRIORITY_S0;
    else if (s1_info->version > s0_info->version)
        priority = SLOT_PRIORITY_S1;
}
```

Commit:
* review comments:
  - fixed potential nullptr dereference
  - printk compile warnings
  - added MCUBOOT_IMAGE_TRAILER_STATUS_UNKNOWN for logging
  - improved logic in slot_priority_from_image_trailers
  - logs slot states during boot in check_and_fix_image_trailer
  - comments updated from review

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -99,26 +106,41 @@ static uint32_t check_and_fix_image_trailer(const char* prefix, const mcuboot_im
+				break; /* SLOT_PRIORITY_S0 */
+			default:
 				result = SLOT_PRIORITY_S1;
+			}
```
Does the compiler not cry about missing default case here?

mads-bn replied:
no - I did not see any warnings related to that, I can add a break but it must be minor

Emil-Juhl replied:
Just usually switch statements require all cases to be covered (or have a default) so I figured if we ever change compiler flags to have more warnings or something, this is a low hanging fruit to fix preemptively.
But in practice it's minor, yeah

mads-bn replied:
we will keep this - at least for now

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +41,120 @@ struct counter_collection {
+
+/* Note1: interface cloned by mcuboot/boot/boot_serial/src/boot_serial.c */
+/* Note2: these constants rely on bits in flash can be changed from 1 to 0 without erasing */
+#define MCUBOOT_IMAGE_TRAILER_STATUS_UNKNOWN  0xFFFFFFFF
```
Not actually sure this ends up used anywhere? (I mean, as opposed to e.g. going with `INACTIVE` as the default/unknown value) but it might be nice in the debug logs tho!

mads-bn replied:
it is only for the debug logs - will keep it

Emil-Juhl commented on subsys/bootloader/bl_storage/bl_storage.c:
```diff
@@ -39,6 +41,120 @@ struct counter_collection {
+ * 3) mcuboot writes mcuboot_image_trailer with value mcuboot_image_trailer_init as final step of image upload
+ * 4) device is rebooted via "mcumgr reboot" or external reset
+ * 5) b0 checks if one or both mcuboot slots have a trailer.magic, if not normal boot proceeds
+ * 6) b0 updates mcuboot_image_trailer.status appropriately:
+ *    From TESTING to BOOTING (if any TESTING)
+ *    From BOOTING to INACTIVE (if any BOOTING)
+ * 7) b0 gives a slot 1st priority if its trailer.status=TESTING
+	  otherwise it prioritizes trailer.status=ACTIVE
+ * 8) mcuboot changes trailer.status of booted image to ACTIVE if confirmed by "mcumgr image confirm"
+ *    and simultaneously marks the not-booted image as INACTIVE
+ * NOTE: if any image fails to validate, the other bank will still be attempted as well.
+ *       Hence b0 can "fallback" to an image otherwise marked as INACTIVE.
```
I think github and indentation was not my friends today ;) Maybe you wanna fix indentation and then we're good to go basically?

mads-bn replied:
I hand picked all your suggestions this time. I tried to make it look good but will look again now

mads-bn replied:
we will keep this - at least for now
